swupd-extract/main.go

```diff
@@ -0,0 +1,516 @@
+			// its contents if not necessary.
+			if srcFI.IsDir() && dstFI.IsDir() {
+				if srcFI.Mode() != dstFI.Mode() {
+					fmt.Printf("! fixing mode for %s from %s to %s\n", dst, dstFI.Mode(), srcFI.Mode())
+					err = os.Chmod(dst, srcFI.Mode())
+					if err != nil {
+						return fmt.Errorf("couldn't fix mode for existing file %s: %s", dst, err)
+					}
+				}
+				srcStat := srcFI.Sys().(*syscall.Stat_t)
+				dstStat := dstFI.Sys().(*syscall.Stat_t)
+				if srcStat.Uid != dstStat.Uid || srcStat.Gid != dstStat.Gid {
+					fmt.Printf("! fixing ownership for %s from %d:%d to %d:%d\n", dst, dstStat.Uid, dstStat.Gid, srcStat.Uid, srcStat.Gid)
+					err = os.Chown(dst, int(srcStat.Uid), int(srcStat.Gid))
+					if err != nil {
+						return fmt.Errorf("couldn't fix ownership of existing file %s: %s", dst, err)
+					}
+				}
+				continue
+			}
+
+			if noOverwrite {
+				fmt.Printf("! skipping %s\n", dst)
+				continue
+			}
+
+			fmt.Printf("! overwriting %s\n", dst)
+			err = os.RemoveAll(dst)
+			if err != nil {
+				return fmt.Errorf("couldn't remove %s for overwriting: %s", dst, err)
+			}
+		}
+
+		switch f.Type {
+		case swupd.TypeFile:
+			err = copyFile(dst, src, srcFI)
+		case swupd.TypeDirectory:
+			err = os.Mkdir(dst, srcFI.Mode())
+			if err != nil {
+				return err
+			}
+			err = os.Chmod(dst, srcFI.Mode())
+		case swupd.TypeLink:
+			var linkname string
+			linkname, err = os.Readlink(src)
+			if err != nil {
+				return err
+			}
+			err = os.Symlink(linkname, dst)
+		default:
+			err = fmt.Errorf("unknown type for %s", f.Name)
+		}
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func copyFile(dst string, src string, srcFI os.FileInfo) error {
+	srcFile, err := os.Open(src)
+	if err != nil {
+		return err
+	}
+
+	dstFile, err := os.OpenFile(dst, os.O_RDWR|os.O_CREATE|os.O_TRUNC, srcFI.Mode())
+	if err != nil {
+		_ = srcFile.Close()
+		return err
+	}
+
+	_, err = io.Copy(dstFile, srcFile)
+	if err != nil {
+		_ = srcFile.Close()
+		_ = dstFile.Close()
+		return fmt.Errorf("failure to copy data from %s to %s: %s", src, dst, err)
+	}
+	err = srcFile.Close()
+	if err != nil {
+		_ = dstFile.Close()
+		return err
+	}
+
+	srcStat := srcFI.Sys().(*syscall.Stat_t)
+	err = dstFile.Chown(int(srcStat.Uid), int(srcStat.Gid))
+	if err != nil {
+		_ = dstFile.Close()
+		return err
+	}
+	if srcFI.Mode()&(os.ModeSticky|os.ModeSetgid|os.ModeSetuid) != 0 {
+		err = dstFile.Chmod(srcFI.Mode())
+		if err != nil {
+			_ = dstFile.Close()
+			return err
+		}
+	}
+	return dstFile.Close()
+}
+
+func verifySignature(content, sig, cert string) error {
+	cmd := exec.Command(
+		"openssl", "smime", "-verify",
+		"-in", sig,
+		"-inform", "der",
+		"-content", content,
+		"-CAfile", cert,
+		"-purpose", "crlsign",
+	)
+	var buf bytes.Buffer
+	cmd.Stdout = &buf
+	cmd.Stderr = &buf
+	err := cmd.Run()
+	if err != nil {
+		return fmt.Errorf("signature verification failed: %s\nOUTPUT:\n%s", err, buf.Bytes())
```

Comment:
Maybe worth outputting the command line used when it fails?

Reply:
Done.